crypto/rsa_extra/rsassa_pss_asn1_test.cc

```diff
@@ -414,20 +417,110 @@ struct PssParamsInvalidInput {
+} kPssConversionTestInputs[] = {
+    {EVP_sha1(), PSS_DEFAULT_SALT_LEN, 0, 1, 1},
+    {EVP_sha224(), 30, 0, 0, 0},
+    {EVP_sha256(), 30, 0, 0, 0},
+    {EVP_sha384(), 30, 0, 0, 0},
+    {EVP_sha512(), 30, 0, 0, 0},
+    // This test case happens in |pkey_rsa_init| of |p_rsa.c|.
+    {nullptr, -2, 1, 1, 1},
+};
+
+static void test_RSASSA_PSS_PARAMS_get(RSASSA_PSS_PARAMS *pss,
+                                       const EVP_MD *expect_md,
+                                       int expect_saltlen) {
+  const EVP_MD *sigmd = nullptr;
+  const EVP_MD *mgf1md = nullptr;
+  int saltlen = -1;
+  ASSERT_TRUE(RSASSA_PSS_PARAMS_get(pss, &sigmd, &mgf1md, &saltlen));
+  EXPECT_EQ(sigmd, expect_md);
+  EXPECT_EQ(mgf1md, expect_md);
+  EXPECT_EQ(saltlen, expect_saltlen);
+}
+
+class PssConversionTest : public testing::TestWithParam<PssConversionTestInput> {};
+
+// This test is to check the conversion between |RSASSA_PSS_PARAMS| and
+// (|*sigmd|, |*mgf1md| and |saltlen|), which are fields of |RSA_PKEY_CTX|.
+// The 1st step is to validate |RSASSA_PSS_PARAMS_create|.
+// The 2nd step is to validate |RSASSA_PSS_PARAMS_get|.
+TEST_P(PssConversionTest, CreationAndGetSuccess) {
+  const auto &param = GetParam();
+  RSASSA_PSS_PARAMS *pss = nullptr;
+  const EVP_MD *md = param.md;
+  int saltlen = param.saltlen;
+  // STEP 1: validate |RSASSA_PSS_PARAMS_create|.
+  ASSERT_TRUE(RSASSA_PSS_PARAMS_create(md, md, saltlen, &pss));
+  // Validate if the pss should be allocated.
+  if (param.expect_pss_is_null) {
+    ASSERT_FALSE(pss);
+    return;
+  } else {
+    ASSERT_TRUE(pss);
+  }
+  // Holds ownership of heap-allocated RSASSA_PSS_PARAMS.
+  bssl::UniquePtr<RSASSA_PSS_PARAMS> pss_ptr(pss);
+  // Validate the hash_algor of pss.
+  if (param.expect_pss_hash_is_null) {
+    // hash_algor is NULL when the default value (sha1) is passed.
+    // Pss params encode expects this default value is omitted.
+    // See 3.1. https://tools.ietf.org/html/rfc4055#section-3.1
+    // MUST omit the hashAlgorithm field when SHA-1 is used.
+    EXPECT_FALSE(pss->hash_algor);
+    EXPECT_FALSE(pss->mask_gen_algor);
+  } else {
+    ASSERT_TRUE(pss->hash_algor);
+    ASSERT_TRUE(pss->mask_gen_algor);
+    ASSERT_TRUE(pss->mask_gen_algor->one_way_hash);
+    EXPECT_EQ(pss->hash_algor->nid, EVP_MD_type(md));
+    EXPECT_EQ(pss->mask_gen_algor->one_way_hash->nid, EVP_MD_type(md));
+  }
+  // Validate the saltlen of pss.
+  if (param.expect_pss_saltlen_is_null) {
+    // salt_len is NULL when the default value (20) is passed.
+    // Pss params encode expects the default value to be omitted.
+    // This is not a MUST but it is implemented in OpenSSL.
+    EXPECT_FALSE(pss->salt_len);
+  } else {
+    ASSERT_TRUE(pss->salt_len);
+    EXPECT_EQ(pss->salt_len->value, saltlen);
+  }
+  // Trailer field should not be set because it's for encoding only.
+  // See 3.1. https://tools.ietf.org/html/rfc4055#section-3.1
+  // MUST omit the trailerField field.
```

Comment:
In general the trailer field is not required to be absent. Only when you perform signature generation. So, I am not sure we can assume this here.

Reply:
* There is no `only` before `perform signature generation`.
* I think the implementation is safe because 
   * both OpenSSL and JDK11 omit the trailerField field.
   * this will be used for encoding pss params. The decode has to accept both absent and 1.
* We discussed this topic before https://github.com/awslabs/aws-lc/pull/89#discussion_r581415082